benchmark/operation_test.go

```diff
@@ -0,0 +1,101 @@
+
+func BenchmarkClientWrite(b *testing.B) {
+	benchmarks := []struct {
+		name string
+		opt  *options.ClientOptions
+	}{
+		{name: "not compressed", opt: options.Client().ApplyURI("mongodb://localhost:27017")},
+		{name: "snappy", opt: options.Client().ApplyURI("mongodb://localhost:27017").SetCompressors([]string{"snappy"})},
+		{name: "zlib", opt: options.Client().ApplyURI("mongodb://localhost:27017").SetCompressors([]string{"zlib"})},
+		{name: "zstd", opt: options.Client().ApplyURI("mongodb://localhost:27017").SetCompressors([]string{"zstd"})},
+	}
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			client, err := mongo.NewClient(bm.opt)
+			if err != nil {
+				b.Fatalf("error connecting: %v", err)
+			}
+			ctx := context.Background()
+			err = client.Connect(ctx)
+			if err != nil {
+				b.Fatalf("error connecting: %v", err)
+			}
+			defer client.Disconnect(context.Background())
+			coll := client.Database("test").Collection("test")
+
+			b.ResetTimer()
+			b.RunParallel(func(p *testing.PB) {
+				for p.Next() {
+					_, err := coll.InsertOne(context.Background(), bson.D{{"text", text}})
```

Comment:
You are right. The `BenchmarkClientBulkWrite` did not cover the loop properly because `len(op.Batches.Current)` is 1.

However, we do not have to extend `dst` after using the memory pool. We recycle the space of `dst`.  Therefore, `dst` will grow to a reasonably big size over time. On the other hand, `totalExpectedSize` = `len(op.Batches.Current)` x `len(doc)`, which can not be obtained easily. Overall, I doubt if it is worth a pre-allocation.